mcs/class/corlib/System/AppDomainSetup.cs

```diff
@@ -139,14 +140,19 @@ static string GetAppBase (string appBase)
 				appBase = appBase.Substring (7);
 				if (Path.DirectorySeparatorChar != '/')
 					appBase = appBase.Replace ('/', Path.DirectorySeparatorChar);
-				if (Environment.IsRunningOnWindows) {
-					// Under Windows prepend "//" to indicate it's a local file
-					appBase = "//" + appBase;
+			}
+			appBase = Path.GetFullPath (appBase);
+
+			if (Path.DirectorySeparatorChar != '/') {
+				bool isExtendedPath = appBase.StartsWith (@"\\?\", StringComparison.Ordinal);
+				if (appBase.IndexOf (':', isExtendedPath ? 6 : 2) != -1) {
+					throw new NotSupportedException ("The given path's format is not supported.");
 				}
-			} else {
-				appBase = Path.GetFullPath (appBase);
 			}
 
+			// Trigger a validation of the path
+			new FileIOPermission (FileIOPermissionAccess.PathDiscovery, appBase);
```

Comment:
Please clarify, I'm not sure what CAS means in this context. The call to the FileIOPermission() constructor triggers a validation of the path, checking for invalid path characters such as ? etc on Windows. If there's a better way to do this using other APIs then please let me know.